posts/2022-09-22-Rust-1.64.0.md

```diff
@@ -0,0 +1,260 @@
+pub struct StorageRequest(bool);
+
+impl StorageRequest {
+    /// Create a new instance of `StorageRequest`.
+    pub fn new() -> Self { ... }
+    /// Decide whether debug mode should be enabled.
+    pub fn set_debug(self, b: bool) -> Self { ... }
+    /// Send the request and receive a response.
+    pub async fn send(self) -> Result<StorageResponse, Error> { ... }
+}
+
+// The new implementations:
+// 1. create a new named future type
+// 2. implement `IntoFuture` for `StorageRequest`
+pub type StorageRequestFuture = Pin<Box<dyn Future<Output = Result<StorageResponse, Error> + Send + 'static>>
+impl IntoFuture for StorageRequest {
+    type IntoFuture = StorageRequestFuture;
+    type Output = <StorageRequestFuture as Future>::Output;
+    fn into_future(self) -> Self::IntoFuture {
+        Box::pin(self.send())
+    }
+}
+```
+
+This takes a bit more code to implement, but provides a simpler API for users.
+
+In the future, the Rust Async WG hopes to simplify the creating new named
+futures by supporting [`impl Trait` in `type` aliases (Type Alias Impl Trait or
+TAIT)](https://rust-lang.github.io/impl-trait-initiative/explainer/tait.html).
+This should make implementing `IntoFuture` easier by simplifying the type
+alias' signature, and make it more performant by removing the `Box` from the
+type alias.
+
+### Stabilizing `ready!` and `poll_fn`
+
+When working with the low-level details of futures, you'll currently often
+encounter the [`Poll`](https://doc.rust-lang.org/std/task/enum.Poll.html) type.
+We've stabilized a few helpers for working with this type:
+
+- [`ready!`](https://doc.rust-lang.org/std/task/macro.ready.html) takes a
+  `Poll` and returns from the containing function for `Poll::Pending`, or
+  returns the contained value in a `Poll::Ready`. This can be useful when
+  implementing futures manually using `Poll`.
+- [`poll_fn`](https://doc.rust-lang.org/std/future/fn.poll_fn.html) turns a
+  function returning `Poll` into a future.
+
+In the future, we hope to provide simpler APIs that require less use of
+low-level details like `Poll` and `Pin`, but in the meantime, these helpers
+make it easier to write such code.
+
+### Minimum required kernel and glibc versions increased
+
+As [previously
+announced](https://blog.rust-lang.org/2022/08/01/Increasing-glibc-kernel-requirements.html),
+`linux` targets now require at least Linux kernel 3.2 (except for targets which
+already required a newer kernel), and `linux-gnu` targets now require glibc
+2.17 (except for targets which already required a newer glibc).
+
+### Profile-guided optimization (PGO) for Windows targets
+
+Windows builds of the Rust compiler now use profile-guided optimization,
+providing performance improvements of 10-20% for compiling Rust code on
+Windows.
+
+### `rust-analyzer` now available via `rustup`, no longer in preview
+
+Users of `rust-analyzer` can now obtain it via `rustup` on the stable and beta
+channels, not just on nightly.
```

Comment:
This seems like a big news to me, and we should make it more prominent and write a tiny more, possibly including the command to add rust-analyzer with rustup.

Reply:
A caveat is that they _don't_ want VS Code users to go that way, as the extension grabs the weekly rust-analyzer builds itself.

Reply:
I was hoping to have a new rustup release ready before 1.64. As it stands now, it is a bit awkward to use.  The user needs to run `rustup which rust-analyzer` to get the path to the executable to run.  (With a new release, it would just be `rust-analyzer`).

I would encourage adding some text here to clarify what this means or what the consequences are. Maybe something along the lines of:

> [rust-analyzer](https://rust-analyzer.github.io/) is now included as part of the collection of tools included with Rust. This can make it easier to download and access rust-analyzer, and makes it available on more platforms. It is available as a [rustup component](https://rust-lang.github.io/rustup/concepts/components.html) which can be installed with `rustup component add rust-analyzer`. 
> 
> At this time, to run the rustup-installed version, use the command `rustup which rust-analyzer` to obtain the path to the executable. The next release of rustup will have a built-in proxy so that running the executable `rust-analyzer` will launch the appropriate version.
> 
> Most users should continue to use the releases provided by the rust-analyzer team (available on the [rust-analyzer releases page](https://github.com/rust-lang/rust-analyzer/releases)), which are published more frequently. Users of the [official VSCode extension](https://marketplace.visualstudio.com/items?itemName=rust-lang.rust-analyzer) are not affected since it automatically downloads new releases.

However, after writing that, it seems like way too many words for something that I think is relatively minor.

Reply:
Note that there is also `rustup run rust-analyzer` which works, even though it's not a singl ebinary.